### ISSUE TITLE
JVM sometimes hangs when calling kill -QUIT <javapid>

### ISSUE THREAD

**Issue by pwagland** [eclipse-openj9/openj9#12224]
Java -version output
--------------------

Output from `java -version`.
```
# java --version
openjdk 11.0.10 2021-01-19
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.10+9)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.24.0, JRE 11 Linux amd64-64-Bit Compressed References 20210120_910 (JIT enabled, AOT enabled)
OpenJ9   - 345e1b09e
OMR      - 741e94ea8
JCL      - 0a86953833 based on jdk-11.0.10+9)
```

Summary of problem
------------------

We have a tomcat instance that randomly hangs. We narrowed this down to when our docker health check is failing, which in turn does a `pkill -3 java` so that we could diagnose _why_ our health checks were getting stuck. However, it appears that sometimes when doing this the JVM just freezes, and nothing happens anymore. `jstack` gives no results.

We can force this to occur with the following:

```
$ while true; do pkill -3 java; done
```

Typically this does within a few hundred thread dumps, although I have had it once generate almost 3000. I have never had this not fail within about 10 minutes.

Diagnostic files
----------------

There is no crash, nor is there any output.

Given that we _can_ reproduce this 100%, we can test and/or retrieve information that might help to diagnose the issue.

**Comment by DanHeidinga** [comment]
fyi @babsingh

**Comment by pwagland** [comment]
FWIW, we are also able to reproduce this behaviour with 11.0.9 as well. I have not tried older versions.

**Comment by babsingh** [comment]
@pwagland `pkill -3 java` sends a `SIGQUIT`, which generates a `javacore*.txt`. The title mentions that the `JVM hangs when calling kill -HUP (-1)`. I have a few questions in order to better understand the hang.
- Can you confirm if the hang happens only when `SIGQUIT` is raised a few 100 times? 
- Does `SIGHUP` terminate the JVM correctly when `SIGQUIT` is not raised?
- Can you generate and provide us **core files after the JVM hangs**? `kill -6 <PID>` (`SIGABRT`) will generate a `core*.dmp`, `javacore*.txt` and few other files before terminating the JVM.

**Comment by pwagland** [comment]
@babsingh - Gah! You are right, it is a `-QUIT`, not a `-HUP`!

1. It is only guaranteed after a few hundred (thousand?) times. I have seen it occur after only 6 occurrences. I have seen it take ~2800 kills to trigger the hang.
2. I have not tried this, I will run the test again and see.
3. I have not tried this, I will run the test again and see.

**Comment by pwagland** [comment]
@babsingh 

1. In production, where we see this in about 4-5% of our docker stacks, this happens with <10 kills.
2. `kill -HUP <pid>` does nothing once java has hung.
3. `kill -6 <pid>` kills the process and creates four files.

I am now trying to see if I can reproduce in a simpler container so that I can upload these files.

**Comment by pwagland** [comment]
With respect to 3. What _is_ interesting, is that this causes two thread dumps to be created, the one that it was hung on, and a new one for shutting down.

```
-rw-r----- 1 12 12    247280 Mar 16 18:46 javacore.20210316.184647.7.0244.txt
-rw------- 1 12 12 641122304 Mar 16 18:48 core.20210316.184837.7.0246.dmp
-rw-r----- 1 12 12    207923 Mar 16 18:48 javacore.20210316.184647.7.0245.txt
-rw-r----- 1 12 12    600485 Mar 16 18:48 javacore.20210316.184837.7.0247.txt
-rw-r----- 1 12 12    395263 Mar 16 18:48 verbosegc.20210316.184527.txt.001
-rw-r----- 1 12 12    316168 Mar 16 18:48 Snap.20210316.184837.7.0248.trc
-rw-r----- 1 12 12       169 Mar 16 18:48 jitdump.20210316.184837.7.0249.dmp
```

There is not javacore `246`, and 245 is the (I suspect anyway) the one that it was trying to generate when it hung. I have attached both javacore files.

I still have the core file, if you need this, then I'll need to get somewhere to upload it, as it is 66M compressed.

[javacore.20210316.184647.7.0245.txt.gz](https://github.com/eclipse/openj9/files/6151587/javacore.20210316.184647.7.0245.txt.gz)
[javacore.20210316.184837.7.0247.txt.gz](https://github.com/eclipse/openj9/files/6151588/javacore.20210316.184837.7.0247.txt.gz)
[jitdump.20210316.184837.7.0249.dmp.gz](https://github.com/eclipse/openj9/files/6151589/jitdump.20210316.184837.7.0249.dmp.gz)
[Snap.20210316.184837.7.0248.trc.gz](https://github.com/eclipse/openj9/files/6151590/Snap.20210316.184837.7.0248.trc.gz)

**Comment by babsingh** [comment]
> I still have the core file, if you need this, then I'll need to get somewhere to upload it, as it is 66M compressed.

The `javacores` don't have function names for the native stack. So, I will need the core file to study the native stack in `gdb`.

https://ibm.ent.box.com/f/6400941f71d8467b89239b383afddbb2: Does this link allow you to upload files? I will also need the `JDK` (`java` executable) to open the core file in `gdb` with symbols.

**Comment by pwagland** [comment]
@babsingh, maybe you already have a notification, but I have uploaded the files to this container. I have included `java` as well, but to the best of my knowledge, this is just https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.10%2B9_openj9-0.24.0/OpenJDK11U-jdk_x64_linux_openj9_11.0.10_9_openj9-0.24.0.tar.gz

**Comment by babsingh** [comment]
@pwagland Thanks for providing the core file. Unfortunately, I am unable to read the native stack in the core file:
```
warning: .dynamic section for "/lib64/ld-linux-x86-64.so.2" is not at the expected address (wrong library or version mismatch?)
```
I will have to view the core file on the same machine where it gets generated. Can you provide me the instructions to reproduce the hang? I will try to generate and view core files on my local machine.

**Comment by pshipton** [comment]
We use `jextract` to obtain the system libraries from the original machine so the core can be viewed on other machines.
https://www.eclipse.org/openj9/docs/tool_jextract/

Also note the debug image is available, containing additional debug symbols.
https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.10%2B9_openj9-0.24.0/OpenJDK11U-debugimage_x64_linux_openj9_11.0.10_9_openj9-0.24.0.tar.gz

**Comment by pwagland** [comment]
@babsingh, @pshipton, I will look at this tomorrow. I tried, but cannot create an open simple test case, which means that it _might_ be due to one of the command line arguments passed in? A different GC setting maybe? Anyway, I can reproduce, and extract the core tomorrow.

Let me know if would be more useful to do this with the debug image, and I will try to reproduce with that, and see what happens.

**Comment by pshipton** [comment]
@pwagland you just need to run jextract on any core file you already have, and it will zip up the system libraries and the jvm libraries, and the core. You could remove the core file itself from the zip if Babneet already has it. The debug image can be added by Babneet before running gdb on the core, to make additional debug symbols available in gdb.

**Comment by pwagland** [comment]
@pshipton thanks for the clarification!

@babsingh I have added core.20210316.184837.7.0246.dmp.zip to the aforementioned share that you gave.

**Comment by ninja-** [comment]
not sure if that's the same issue, but when I get OutOfMemory error and openj9 starts preparing the dumps of everything, very often it hangs. java 15 version. it seems like a regression made in last few months...

**Comment by ninja-** [comment]
it's easy to reproduce, but I am not sure gdb can provide something useful.

```
Thread 3 (LWP 80 "Signal Reporter"):
#0  0x00007ff5a82b04d4 in ?? () from /usr/glibc-compat/lib/libpthread.so.0
#1  0x00007ff5a82b05c8 in ?? () from /usr/glibc-compat/lib/libpthread.so.0
#2  0x00007ff5a6ec23a8 in asynchSignalReporter () from /usr/lib/jvm/openjdk/lib/compressedrefs/libj9prt29.so
#3  0x00007ff5a733b432 in thread_wrapper () from /usr/lib/jvm/openjdk/lib/compressedrefs/libj9thr29.so
#4  0x00007ff5a82a6f00 in ?? () from /usr/glibc-compat/lib/libpthread.so.0
#5  0x00007ff5a81d245f in clone () from /usr/glibc-compat/lib/libc.so.6

Thread 2 (LWP 79 "DestroyJavaVM h"):
#0  0x00007ff5a82ada7c in pthread_cond_wait () from /usr/glibc-compat/lib/libpthread.so.0
#1  0x00007ff5a7335927 in omrthread_monitor_wait () from /usr/lib/jvm/openjdk/lib/compressedrefs/libj9thr29.so
#2  0x00007ff5a75ee314 in protectedDestroyJavaVM () from /usr/lib/jvm/openjdk/lib/compressedrefs/libj9vm29.so
#3  0x00007ff5a6ec2da3 in omrsig_protect () from /usr/lib/jvm/openjdk/lib/compressedrefs/libj9prt29.so
#4  0x00007ff5a75ed8a9 in DestroyJavaVM () from /usr/lib/jvm/openjdk/lib/compressedrefs/libj9vm29.so
#5  0x00007ff5a7af924a in DestroyJavaVM () from /usr/lib/jvm/openjdk/lib/compressedrefs/libjvm.so
#6  0x00007ff5a82c4de4 in JavaMain () from /usr/lib/jvm/openjdk/bin/../lib/libjli.so
#7  0x00007ff5a82c8769 in ThreadJavaMain () from /usr/lib/jvm/openjdk/bin/../lib/libjli.so
#8  0x00007ff5a82a6f00 in ?? () from /usr/glibc-compat/lib/libpthread.so.0
#9  0x00007ff5a81d245f in clone () from /usr/glibc-compat/lib/libc.so.6

Thread 1 (LWP 77 "java"):
#0  0x00007ff5a82a84e7 in ?? () from /usr/glibc-compat/lib/libpthread.so.0
#1  0x00007ff5a82c93bd in CallJavaMainInNewThread () from /usr/lib/jvm/openjdk/bin/../lib/libjli.so
#2  0x00007ff5a82c627d in ContinueInNewThread () from /usr/lib/jvm/openjdk/bin/../lib/libjli.so
#3  0x00007ff5a82c791b in JLI_Launch () from /usr/lib/jvm/openjdk/bin/../lib/libjli.so
#4  0x000055add4c9da83 in main ()
```

```
2021/03/17 18:17:03 JVMDUMP039I Processing dump event "systhrow", detail "java/lang/OutOfMemoryError" at 2021/03/17 18:17:03 - please wait.
2021/03/17 18:17:03 JVMDUMP032I JVM requested System dump using '/app/core.20210317.181703.77.0001.dmp' in response to an event
2021/03/17 18:17:04 JVMDUMP010I System dump written to /app/core.20210317.181703.77.0001.dmp
2021/03/17 18:17:04 JVMDUMP032I JVM requested Heap dump using '/app/heapdump.20210317.181703.77.0002.phd' in response to an event
2021/03/17 18:17:05 JVMDUMP010I Heap dump written to /app/heapdump.20210317.181703.77.0002.phd
2021/03/17 18:17:05 JVMDUMP032I JVM requested Java dump using '/app/javacore.20210317.181703.77.0003.txt' in response to an event
2021/03/17 18:17:05 JVMDUMP010I Java dump written to /app/javacore.20210317.181703.77.0003.txt
2021/03/17 18:17:05 JVMDUMP032I JVM requested Snap dump using '/app/Snap.20210317.181703.77.0004.trc' in response to an event
2021/03/17 18:17:05 JVMDUMP010I Snap dump written to /app/Snap.20210317.181703.77.0004.trc
2021/03/17 18:17:05 JVMDUMP046I VM is shutting down. Reason: java/lang/OutOfMemoryError
2021/03/17 18:17:05 JVMDUMP039I Processing dump event "systhrow", detail "java/lang/OutOfMemoryError" at 2021/03/17 18:17:05 - please wait.
2021/03/17 18:17:06 JVMDUMP039I Processing dump event "systhrow", detail "java/lang/OutOfMemoryError" at 2021/03/17 18:17:06 - please wait.
2021/03/17 18:17:06 JVMDUMP032I JVM requested Heap dump using '/app/heapdump.20210317.181706.77.0007.phd' in response to an event
2021/03/17 18:17:07 JVMDUMP010I Heap dump written to /app/heapdump.20210317.181706.77.0007.phd
2021/03/17 18:17:07 JVMDUMP032I JVM requested Java dump using '/app/javacore.20210317.181706.77.0008.txt' in response to an event
2021/03/17 18:17:07 JVMDUMP010I Java dump written to /app/javacore.20210317.181706.77.0008.txt
2021/03/17 18:17:07 JVMDUMP032I JVM requested Snap dump using '/app/Snap.20210317.181706.77.0009.trc' in response to an event
2021/03/17 18:17:07 JVMDUMP010I Snap dump written to /app/Snap.20210317.181706.77.0009.trc
2021/03/17 18:17:07 JVMDUMP046I VM is shutting down. Reason: java/lang/OutOfMemoryError
2021/03/17 18:17:10 JVMDUMP039I Processing dump event "systhrow", detail "java/lang/OutOfMemoryError" at 2021/03/17 18:17:10 - please wait.
```

**Comment by ninja-** [comment]
based on logs, maybe the process of dumping does not have a proper lock and two are triggered at a time, deadlocking each other?

**Comment by babsingh** [comment]
> I have added core.20210316.184837.7.0246.dmp.zip to the aforementioned share that you gave.

I was able to read the core file. But, it contained native stack for only one thread. `jdmpview` showed 38 threads. This core file was created using `SIGABRT`. @pshipton How to get native stack for all threads? Will `SIGSEGV` (`kill -11 <PID>`) create a core file with native stack for all threads? or Do we need to live debug to acquire native stack for all threads (`gdb jdk11/bin/java <PID>`)?

Also, I couldn't recreate the hang using a simple `Object.wait()` micro-test, which was sent `SIGQUIT` 1000 times in 0.5 second intervals.

```
(gdb) info threads
  Id   Target Id                         Frame
* 1    Thread 0x7f3148e38700 (LWP 81251) __pthread_kill (threadid=<optimized out>, signo=11) at ../sysdeps/unix/sysv/linux/pthread_kill.c:57

(gdb) where
#0  __pthread_kill (threadid=<optimized out>, signo=11) at ../sysdeps/unix/sysv/linux/pthread_kill.c:57
#1  0x00007f316a515c7d in omrdump_create () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9prt29.so
#2  0x00007f3169b7c212 in doSystemDump () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9dmp29.so
#3  0x00007f3169b78005 in protectedDumpFunction () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9dmp29.so
#4  0x00007f316a517723 in omrsig_protect () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9prt29.so
#5  0x00007f3169b7b67b in runDumpFunction () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9dmp29.so
#6  0x00007f3169b7b80f in runDumpAgent () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9dmp29.so
#7  0x00007f3169b9375b in triggerDumpAgents () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9dmp29.so
#8  0x00007f3169b7d8fe in abortHandler () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9dmp29.so
#9  <signal handler called>
#10 0x00007f316a516520 in mainASynchSignalHandler () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9prt29.so
#11 <signal handler called>
#12 0x00007f316cccaad1 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f31640f6e08) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#13 __pthread_cond_wait_common (abstime=0x0, mutex=0x7f3134000d18, cond=0x7f31640f6de0) at pthread_cond_wait.c:502
#14 __pthread_cond_wait (cond=0x7f31640f6de0, mutex=0x7f3134000d18) at pthread_cond_wait.c:655
#15 0x00007f316a2da897 in omrthread_monitor_wait () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9thr29.so
#16 0x00007f3168450758 in TR::CompilationInfoPerThread::doSuspend() () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9jit29.so
#17 0x00007f3168460ed8 in TR::CompilationInfoPerThread::run() () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9jit29.so
#18 0x00007f3168460f6a in protectedCompilationThreadProc(J9PortLibrary*, TR::CompilationInfoPerThread*) () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9jit29.so
#19 0x00007f316a517723 in omrsig_protect () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9prt29.so
#20 0x00007f31684613c4 in compilationThreadProc(void*) () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9jit29.so
#21 0x00007f316a2e03c6 in thread_wrapper () from /root/sigquit_hang/core_file/opt/java/openjdk/lib/compressedrefs/libj9thr29.so
#22 0x00007f316ccc46db in start_thread (arg=0x7f3148e38700) at pthread_create.c:463
#23 0x00007f316c5d871f in clone () from /root/sigquit_hang/core_file/lib/x86_64-linux-gnu/libc.so.6
```

**Comment by ninja-** [comment]
@babsingh easiest repro I think is to use +ExitOnOutOfMemoryError and make code that would trigger out of memory more than one time

**Comment by babsingh** [comment]
> not sure if that's the same issue

More diagnostic information is required to confirm that the `SIGQUIT` and `+ExitOnOutOfMemoryError` hangs are the same. Until then, we should assume that they are different.

> when I get OutOfMemory error and openj9 starts preparing the dumps of everything, very often it hangs forever.

`DestroyJavaVM` thread is waiting on a `mutex` to be notified: https://github.com/eclipse/openj9/blob/1eaf33671f61004ff201e88f687e935424b430dc/runtime/vm/jniinv.c#L320

If only the three threads shown in https://github.com/eclipse/openj9/issues/12224#issuecomment-801809036 are remaining in the JVM, then the `DestroyJavaVM` thread won't receive a notification, and the JVM would stay suspended during termination. 

> +ExitOnOutOfMemoryError and make code that would trigger out of memory more than one time

`+ExitOnOutOfMemoryError` hangs were seen in the past: https://github.com/eclipse/openj9/issues/8552. ++Owner: @theresa-m, for insights. 

++@gacholio: for insights + updated `DestroyJavaVM` last.

**Comment by pshipton** [comment]
> How to get native stack for all threads?

When it's hung, use `gcore <pid>` to generate the core file. This won't take exclusive access and the GC/VM/JIT structures may not be in a good state for reading java stacks, object heap, etc, but all the native stacks should be available.

**Comment by pshipton** [comment]
I suggest using gcore, and then wait one minute and then use gcore again to create a second core, so we can compare the thread states between them.

**Comment by gacholio** [comment]
The destroy is waiting for the current thread to be the only non-daemon thread left in the VM. This is spec behaviour.

**Comment by babsingh** [comment]
@pwagland @ninja- Requesting two core files for the `SIGQUIT` and `+ExitOnOutOfMemoryError` hangs using **`gcore`** (https://github.com/eclipse/openj9/issues/12224#issuecomment-802832341, https://github.com/eclipse/openj9/issues/12224#issuecomment-802833337) and processed with **`jextract`** (https://github.com/eclipse/openj9/issues/12224#issuecomment-800639061, https://github.com/eclipse/openj9/issues/12224#issuecomment-801043170).

**Comment by pshipton** [comment]
Note only one jextract is required per machine/JVM. If you've already run jextract on the same machine with the same JVM and provided those files, we don't need a second copy.

**Comment by pshipton** [comment]
Well, maybe if you were running a different workload then different libraries could be referenced that weren't already collected by the earlier jextract.

**Comment by pwagland** [comment]
@babsingh, @pshipton, where do we find gcore?

It isn't in the binaries downloaded from: https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.10%2B9_openj9-0.24.0/OpenJDK11U-jdk_x64_linux_openj9_11.0.10_9_openj9-0.24.0.tar.gz , nor is it in the debug build: https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.10%2B9_openj9-0.24.0/OpenJDK11U-debugimage_x64_linux_openj9_11.0.10_9_openj9-0.24.0.tar.gz

**Comment by gacholio** [comment]
@pwagland I suspect it's part of the gdb package (it invokes gdb to create the cores).

**Comment by pshipton** [comment]
Ya sorry, it's a Linux tool, not a Java tool.
https://man7.org/linux/man-pages/man1/gcore.1.html

**Comment by pwagland** [comment]
I have uploaded a number of core files from a reproduction that I made today. Five core files with about 10-15s between them. I ran `jextract` on the first core file, and attached that zip here as well. Names are all similar to `core.20210322.112954.8.gz`.

**Comment by babsingh** [comment]
I looked at the core files. All five core files have the same backtrace for all threads. I believe that the hang happens due to `-Xgc:concurrentScavenge`. 

@pwagland Can you remove `-Xgc:concurrentScavenge` or try a different GC policy to see if it resolves the hang? OpenJ9 GC policies: https://www.eclipse.org/openj9/docs/xgcpolicy/. 

@amicic Can you confirm if the below analysis is correct?

### Java options
`JAVA_OPTS=-Dserver1 -Djava.security.egd=file:/dev/./urandom -server -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/opt/redwood/j2ee/cluster/server1/logs -Djava.awt.headless=true -Dlogging.loggerfactory=com.redwood.scheduler.infrastructure.logging.Log4JLoggerFactory  -Dcom.ibm.enableClassCaching=false -Dcom.redwood.scheduler.db.selfupgrade=true                  -Dcom.redwood.scheduler.configuration.boot.cluster.hostId='docker'                  -Dcom.redwood.scheduler.boot.cluster.'docker'.1.bindAddress='docker'                  -Dcom.redwood.scheduler.boot.car.apps-location='/opt/redwood/addons/apps'                  -verbose:sizes                  -Xgc:concurrentScavenge                  -XX:MinRAMPercentage=20                  -XX:IdleTuningMinIdleWaitTime=10                  -XX:IdleTuningMinFreeHeapOnIdle=10                  -Xjit:waitTimeToEnterDeepIdleMode=10000                  -Xtune:virtualized                  -Xjit:enableSelfTuningScratchMemoryUsageBeforeCompile                  -DRP_CONTEXTPATH=                  -DRP_UNPACKWAR='true' -verbose:gc                    -Xverbosegclog:/opt/redwood/j2ee/cluster/server1/logs/verbosegc.%Y%m%d.%H%M%S.txt,10,50000 -XX:MaxRAMPercentage=58.3789 -Dlog4j2.disableJmx=true -Djdk.tls.ephemeralDHKeySize=2048 -Djava.protocol.handler.pkgs=org.apache.catalina.webresources -Dorg.apache.catalina.security.SecurityListener.UMASK=0027`

### Diagnostic Information

```
for all threads,
   0x970: UDATA safePointCount = 0x0000000000000000 (0)

in j9javavm,
   0x5020: UDATA safePointState = 0x0000000000000000 (0)
   0x5028: UDATA safePointResponseCount = 0x0000000000000000 (0)

!stack 0x00018100 !j9vmthread 0x00018100  !j9thread 0x7f922c01c060   tid 0xf3 (243) // (main)
   
   0x98: UDATA publicFlags = 0x00000000000000A1 (161) // Owns VM Access but inNative = 1 
   0x938: UDATA inNative = 0x0000000000000001 (1)

   Native stack trace (waits on a socket):
      #0  0x00007f9230d9acb9 in poll () from /team/babsing/projects/sigquit_hang/libs1/lib/x86_64-linux-gnu/libc.so.6
      #1  0x00007f920c495359 in NET_Timeout () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/libnet.so
      #2  0x00007f920c492d64 in Java_java_net_PlainSocketImpl_socketAccept () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/libnet.so

!stack 0x00171d00 !j9vmthread 0x00171d00  !j9thread 0x7f922c3b2dd8   tid 0x102 (258) // (Dedicated GC Main)
   0x98: UDATA publicFlags = 0x0000000000000021 (33) // Owns VM Access

   Native stack trace:
         #0  0x00007f9231499fb9 in pthread_cond_timedwait@@GLIBC_2.3.2 () from /team/babsing/projects/sigquit_hang/libs1/lib/x86_64-linux-gnu/libpthread.so.0
         #1  0x00007f922aa7ec92 in omrthread_monitor_wait_timed () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9thr29.so
         #2  0x00007f922870b454 in MM_Scavenger::shouldDoFinalNotify(MM_EnvironmentStandard*) () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9gc29.so
         #3  0x00007f9228710281 in MM_Scavenger::getNextScanCache(MM_EnvironmentStandard*) () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9gc29.so
         #4  0x00007f9228734688 in MM_Scavenger::completeScan(MM_EnvironmentStandard*) () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9gc29.so
         #5  0x00007f922873b84d in MM_Scavenger::workThreadScan(MM_EnvironmentStandard*) () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9gc29.so
         #6  0x00007f92286d757b in MM_ParallelDispatcher::run(MM_EnvironmentBase*, MM_Task*, unsigned long) () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9gc29.so
         #7  0x00007f922870aabf in MM_Scavenger::mainThreadConcurrentCollect(MM_EnvironmentBase*) () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9gc29.so
         #8  0x00007f92286c5095 in MM_MainGCThread::handleConcurrent(MM_EnvironmentBase*) () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9gc29.so
         #9  0x00007f92286c5299 in MM_MainGCThread::mainThreadEntryPoint() () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9gc29.so
         #10 0x00007f92286c53cc in MM_MainGCThread::main_thread_proc2(OMRPortLibrary*, void*) () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9gc29.so
         #11 0x00007f922acb7723 in omrsig_protect () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9prt29.so
         #12 0x00007f92286c49f9 in MM_MainGCThread::main_thread_proc(void*) () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9gc29.so
         #13 0x00007f922aa803c6 in thread_wrapper () from /team/babsing/projects/sigquit_hang/libs1/opt/java/openjdk/lib/compressedrefs/libj9thr29.so
         #14 0x00007f92314936db in start_thread () from /team/babsing/projects/sigquit_hang/libs1/lib/x86_64-linux-gnu/libpthread.so.0
         #15 0x00007f9230da771f in clone () from /team/babsing/projects/sigquit_hang/libs1/lib/x86_64-linux-gnu/libc.so.6

Other VM Access thread states,
   publicFlags = 0x0000000000000081 (129)
   publicFlags = 0x0000000000000089 (137)
   publicFlags = 0x0000000000020081 (131201)
   publicFlags = 0x0000000000020089 (131209)

Threads have halted for J9_PUBLIC_FLAGS_HALT_THREAD_EXCLUSIVE.

=== VM Access Public Flags ===
   #define J9_PUBLIC_FLAGS_HALT_THREAD_EXCLUSIVE 0x1
   #define J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS 0x8
   #define J9_PUBLIC_FLAGS_VM_ACCESS 0x20
   #define J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE 0x80
   #define J9_PUBLIC_FLAGS_THREAD_PARKED 0x20000
```

### Analysis

Across the five core files, the `Dedicated GC Main (258)` keeps invoking `omrthread_monitor_wait_timed` from `completeScan -> getNextScanCache -> shouldDoFinalNotify` with a 1 millisecond timeout. It seems to be stuck in a recursive loop. It also owns VM Access. All other threads have halted for `THREAD_EXCLUSIVE`. Until `Dedicated GC Main` releases VM Access, the JVM would stay hung.

### Suspect Code

- `MM_Scavenger::completeScan`: https://github.com/eclipse/omr/blob/165a4a81709504806f14bd8e540d97962cafe5f5/gc/base/standard/Scavenger.cpp#L2483
- `MM_Scavenger::getNextScanCache`: https://github.com/eclipse/omr/blob/165a4a81709504806f14bd8e540d97962cafe5f5/gc/base/standard/Scavenger.cpp#L2304
- `MM_Scavenger::shouldDoFinalNotify`: https://github.com/eclipse/omr/blob/165a4a81709504806f14bd8e540d97962cafe5f5/gc/base/standard/Scavenger.cpp#L2208

**Comment by amicic** [comment]
This is a correct analysis.
That GC thread is supposed to yield and release VM access, if there is an external exclusive VM access request (such as generation of javacore). To yield, we look both at thread's HALT_EXCLUSIVE being set (seems to be the case) and exclusiveAccessState being PENDING (please check). If indeed are both set, I don't know why we are still looping - needs to be debugged...

**Comment by amicic** [comment]
As mentioned, removing -Xgc:concurrentScavenge will likely resolve the problem.

**Comment by gacholio** [comment]
I wonder if we hit a timing hole (or bad assumption) related to handoff.

**Comment by gacholio** [comment]
What is the value of exclusiveAccessState? I don't think safe point is involved here.

**Comment by babsingh** [comment]
> What is the value of exclusiveAccessState?

`exclusiveAccessState` is set to `J9_XACCESS_EXCLUSIVE` (`== 2`) across all five core files.
```
   0x1ae0: UDATA exclusiveAccessState = 0x0000000000000002 (2)
```

**Comment by gacholio** [comment]
Which means someone is holding exclusive already - presumably the GC master thread (the only one in the VM with VM access).

**Comment by gacholio** [comment]
The thread holding exclusive will have `vmThread->omrVMThread->exclusiveCount != 0`. Which thread is that?

**Comment by amicic** [comment]
> Which means someone is holding exclusive already - presumably the GC master thread (the only one in the VM with VM access).


Yeah, that's quite possible. The stack trace indicates that we think we are running concurrently, while we actually might be in STW (holding exclusive). That would explain why we are looping, but still need to understand how we may end up in this concurrent/STW mixed scenario...

**Comment by babsingh** [comment]
> The thread holding exclusive will have vmThread->omrVMThread->exclusiveCount != 0. Which thread is that?

Set to 0 for all threads.

**Comment by gacholio** [comment]
I think that means that `acquireExclusiveVMAccessFromExternalThread` was used to acquire it (seems metronome only), the GC is internally manipulating the field or someone is releasing exclusive (the count is decremented before the release takes place).

**Comment by gacholio** [comment]
The rasdump code is potentially using the external thread exclusive. That certainly may be involved here.

**Comment by amicic** [comment]
... or it could be that the thread that acquired exclusive for GC terminated before releasing it?

@pwagland could we get verbose GC logs (matching the core files provided)?

**Comment by gacholio** [comment]
Yes, wondering if the SIGQUIT thread was attached, got exlusive via external, and ended before releasing. This is the only way I can explain every thread having the HALT_EXCLUSIVE bit set.

**Comment by pwagland** [comment]
@amicic I can probably do one better. Hereby my reproduction case:

You need three terminals, then in terminal:
1. Run `docker run --name github_12224_test --env JAVA_TOOL_OPTIONS='-Xgc:concurrentScavenge -XX:+IgnoreUnrecognizedVMOptions -XX:+IdleTuningGcOnIdle -Xshareclasses:name=openj9_system_scc,cacheDir=/opt/java/.scc,readonly,nonFatal' -ti --rm tomcat:9.0.44-jdk11-adoptopenjdk-openj9`
2. Run `docker exec -ti github_12224_test bash -c 'watch -d ls -lt'`
3. Run `docker exec -ti github_12224_test bash -c 'while true; do pkill -3 java; done'`

Once the output from terminal 2 stops, the hang has been hit. For me this typically happens really quickly, between 2-40 core dumps get created before it hangs.

I tried using this as a basis for my prior reproduction case, however I didn't have the `-Xgc:concurrentScavenge` flag, and without that flag, it doesn't hang, or at the very least it produces thousands of thread dumps, instead of very few. So with this test, I can also confirm that `-Xgc:concurrentScavenge` is the most likely suspect, or, at the very least, does something which causes something else to fail.

**Comment by gacholio** [comment]
@pwagland Are the requested dumps generated before the hang?

**Comment by pwagland** [comment]
@gacholio yes. In my three tests I got to see between 2 and 41 dumps before the hang. Without `-Xgc:concurrentScavenge` I let it get up to 1124 dumps before I decided it was probably going to continue working.

**Comment by gacholio** [comment]
Sorry, my question wasn't clear (and you may not be able to answer it easily) - in the case when the hang occurs, are the dumps generated first (presumably you get lots of dumps from non-hang situations)?

**Comment by pwagland** [comment]
@gacholio, right. When the hang occurs, you no longer get any dumps. If I do a `kill -11` at that point, then I get two additional dumps. The one that it appears to have hung on, and the one from the `kill -11`.

**Comment by pwagland** [comment]
@gacholio - right and now I think that the answer is _probably_ that the when the hang occurs, it occurs before the dump is generated. This is based on the fact that when I do the `kill -6` then whatever lock is taken appears to be released, and I suspect that is why I end up getting two additional thread dumps when the `kill -6` happens.

In my previous comment, you can see that I got:
```
-rw-r----- 1 12 12    247280 Mar 16 18:46 javacore.20210316.184647.7.0244.txt
-rw------- 1 12 12 641122304 Mar 16 18:48 core.20210316.184837.7.0246.dmp
-rw-r----- 1 12 12    207923 Mar 16 18:48 javacore.20210316.184647.7.0245.txt
-rw-r----- 1 12 12    600485 Mar 16 18:48 javacore.20210316.184837.7.0247.txt
-rw-r----- 1 12 12    395263 Mar 16 18:48 verbosegc.20210316.184527.txt.001
-rw-r----- 1 12 12    316168 Mar 16 18:48 Snap.20210316.184837.7.0248.trc
-rw-r----- 1 12 12       169 Mar 16 18:48 jitdump.20210316.184837.7.0249.dmp
```

So you can see that the "245" dump turns up two minutes later than the file timestamp indicates that it should. This is the "extra" javacore that I am referring to.

**Comment by gacholio** [comment]
Once we're able to reproduce this here, we can try instrumenting the VM to see if the external case gets used.

**Comment by pwagland** [comment]
I hope that the docker test helps the reproduction! Please let me know if there is anything I can do to help further.

**Comment by pwagland** [comment]
Has there been any progress on this issue? Is there anything that we can do to help diagnose what is going wrong?

**Comment by amicic** [comment]
I had tried to reproduce it, albeit not with docker and tomcat, but still with fairly frequent GC cycles (that would exercise MM_Scavenger::shouldDoFinalNotify path), running for a few hours generating a few thousands for javacores but no luck.
I guess, we to try harder to reproduce it...

**Comment by ninja-** [comment]
@amcic don't you want to start with my much easier variant with ExitOnOutOfMemory?

**Comment by pwagland** [comment]
@amicic were these tests also with `-Xgc:concurrentScavenge` flag, as in my testing this was required to make it fail… note that I have not tested this outside of docker, as that is a more complicated environment for me to setup.

**Comment by amicic** [comment]
Still no luck to reproduce this, but based on analysis around stack trace seen from https://github.com/eclipse/openj9/issues/12224#issuecomment-805125591 I have made an attempt to fix Concurrent Scavenger aspect of it (but could be more than just that).

I have a custom build based on recent head, if you want to give it a try: 
https://ibm.box.com/s/jilhvm2laj5fmw5h6kqv7sooknktkfso

**Comment by pwagland** [comment]
Some good news for the weekend.

I downloaded the java, and it gives:
```
root@5e03a6847fc7:/usr/local/tomcat# java --version
openjdk 11.0.11-internal 2021-04-20
OpenJDK Runtime Environment (build 11.0.11-internal+0-adhoc.jenkins.BuildJDK11x86-64linuxPersonal)
Eclipse OpenJ9 VM (build master-19fa4dcef01, JRE 11 Linux amd64-64-Bit Compressed References 20210416_3027 (JIT enabled, AOT enabled)
OpenJ9   - 19fa4dcef01
OMR      - 5c5e5440b04
JCL      - ba1caf576ee based on jdk-11.0.11+8)
```

Running the test in https://github.com/eclipse/openj9/issues/12224#issuecomment-805239891 I was able to reproduce the issue with the old java very quickly. With the new java I am up to 500 core files… I will let this run overnight, but at least the initial few minutes are very promising!

**Comment by pwagland** [comment]
I left it running overnight, and it made 222889 dumps, and was still going strong, so whatever changes it was that you made they appear to have helped fix this issue, or at least make it non-reproducible in my test. Just for kicks I reran the old test again, and it failed after only 24 dumps, so it's not quite a factor of 10,000, but it's pretty close! I'm happy enough to say that at least this reproduction case is eliminated, or at worst made much more unlikely, by a significant margin.

**Comment by amicic** [comment]
The change (https://github.com/eclipse/omr/pull/5925) has been merged. It has yet to pass nightly regression testing, which may take a couple of days. If does pass, it will  shows up in a nightly build at https://adoptopenjdk.net/. I can let you know when, but of course, the safest is to wait for the next release.

**Comment by pwagland** [comment]
@amicic Thanks!

I have another issue that I still need to create, where we randomly GPF crashes after about a week when allocating memory.  But once that is also added (and resolved 😎) we’ll look forward to upgrading!

**Comment by pshipton** [comment]
Seems this was fixed so closing.
